share/dkg/pedersen/dkg.go

```diff
@@ -1,143 +1,232 @@
-// Package dkg implements the protocol described in "A threshold cryptosystem without a trusted party"
-// by Torben Pryds Pedersen. https://dl.acm.org/citation.cfm?id=1754929.
+// Package dkg implements a general distributed key generation (DKG) framework. This
+// package servers two functionality: (1) to run a fresh new DKG from scratch
+// and (2) to reshare old shares to a potentially distinct new set of nodes (the
+// "resharing" protocol). The
+// former protocol is described in "A threshold cryptosystem without a trusted
+// party" by Torben Pryds Pedersen. https://dl.acm.org/citation.cfm?id=1754929.
+// The latter protocol is implemented in "Verifiable Secret Redistribution for
+// Threshold Signing Schemes", by T. Wong et
+// al.(https://www.cs.cmu.edu/~wing/publications/Wong-Wing02b.pdf)
 package dkg
 
 import (
+	"bytes"
 	"errors"
 
 	"github.com/dedis/kyber"
 
 	"github.com/dedis/kyber/share"
 	vss "github.com/dedis/kyber/share/vss/pedersen"
+	"github.com/dedis/kyber/sign/schnorr"
 )
 
 // Suite wraps the functionalities needed by the dkg package
 type Suite vss.Suite
 
-// DistKeyShare holds the share of a distributed key for a participant.
-type DistKeyShare struct {
-	// Coefficients of the public polynomial holding the public key.
-	Commits []kyber.Point
-	// Share of the distributed secret which is private information.
-	Share *share.PriShare
-	// Coefficients of the private polynomial generated by the node holding the
-	// share. The final distributed polynomial is the sum of all these
-	// individual polynomials, but it is never computed.
-	PrivatePoly []kyber.Scalar
+// Config holds all required information to run a fresh DKG protocol or a
+// resharing protocol. In the case of a new fresh DKG protocol, one must fill
+// the following fields: Suite, Longterm, NewNodes, Threshold (opt). In the case
+// of a resharing protocol, one must fill the following: Suite, Longterm,
+// OldNodes, NewNodes. If the node using this config is creating new shares
+// (i.e. it belongs to the current group), the Share field must be filled in
+// with the current share of the node. If the node using this config is a new
+// addition and thus has no current share, the PublicCoeffs field be must be
+// filled in.
+type Config struct {
+	Suite Suite
+
+	// Longterm is the longterm secret key.
+	Longterm kyber.Scalar
+
+	// Current group of share holders. It can be nil for new DKG. These nodes
```

Comment:
```
	// Current group of share holders. It will be nil for new DKG. These nodes
	// will have invalid shares after the protocol has been run. To be able to issue
	// new shares to a new group, the group member's public key must be inside this
	// list and in the Share field. Keys can be disjoint or not with respect to the
        // NewNodes list.
```